drivers/ethernet/eth_esp32.c

```diff
@@ -0,0 +1,290 @@
+};
+
+static enum ethernet_hw_caps eth_esp32_caps(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
+}
+
+static int eth_esp32_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct eth_esp32_dev_data *dev_data = dev->data;
+	size_t len = net_pkt_get_len(pkt);
+
+	if (net_pkt_read(pkt, dev_data->txb, len)) {
+		return -EIO;
+	}
+
+	uint32_t sent_len = emac_hal_transmit_frame(&dev_data->hal, dev_data->txb, len);
+
+	int res = len == sent_len ? 0 : -EIO;
+
+	return res;
+}
+
+static struct net_pkt *eth_esp32_rx(
+	struct eth_esp32_dev_data *const dev_data, uint32_t *frames_remaining)
+{
+	uint32_t free_rx_descriptor;
+	uint32_t receive_len = emac_hal_receive_frame(
+		&dev_data->hal, dev_data->rxb, sizeof(dev_data->rxb),
+		frames_remaining, &free_rx_descriptor);
+	if (receive_len == 0) {
+		/* Nothing to receive */
+		return NULL;
+	}
+
+	struct net_pkt *pkt = net_pkt_rx_alloc_with_buffer(
+		dev_data->iface, receive_len, AF_UNSPEC, 0, K_MSEC(100));
+	if (pkt == NULL) {
+		eth_stats_update_errors_rx(ctx->iface);
+		LOG_ERR("Could not allocate rx buffer");
+		return NULL;
+	}
+
+	if (net_pkt_write(pkt, dev_data->rxb, receive_len) != 0) {
+		LOG_ERR("Unable to write frame into the pkt");
+		eth_stats_update_errors_rx(ctx->iface);
+		net_pkt_unref(pkt);
+		return NULL;
+	}
+
+	return pkt;
+}
+
+FUNC_NORETURN static void eth_esp32_rx_thread(void *arg1, void *arg2, void *arg3)
+{
+	const struct device *dev = arg1;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (true) {
+		k_sem_take(&dev_data->int_sem, K_FOREVER);
+
+		uint32_t frames_remaining;
+
+		do {
+			struct net_pkt *pkt = eth_esp32_rx(
+				dev_data, &frames_remaining);
+			if (pkt == NULL) {
+				break;
+			}
+
+			if (net_recv_data(dev_data->iface, pkt) < 0) {
+				/* Upper layers are not ready to receive packets */
+				net_pkt_unref(pkt);
+			}
+		} while (frames_remaining > 0);
+	}
+}
+
+IRAM_ATTR static void eth_esp32_isr(void *arg)
+{
+	const struct device *dev = arg;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+	uint32_t intr_stat = emac_ll_get_intr_status(dev_data->hal.dma_regs);
+
+	emac_ll_clear_corresponding_intr(dev_data->hal.dma_regs, intr_stat);
+
+	if (intr_stat & EMAC_LL_DMA_RECEIVE_FINISH_INTR) {
+		k_sem_give(&dev_data->int_sem);
+	}
+}
+
+static int generate_mac(uint8_t mac_addr[6])
+{
+	int res = 0;
+#if DT_INST_PROP(0, zephyr_random_mac_address)
+	gen_random_mac(mac_addr, 0x24, 0xD7, 0xEB);
+#elif !NODE_HAS_VALID_MAC_ADDR(DT_DRV_INST(0))
+	if (esp_read_mac(mac_addr, ESP_MAC_ETH) != ESP_OK) {
+		res = -EIO;
+	}
+#endif
+	return res;
+}
+
+int eth_esp32_initialize(const struct device *dev)
+{
+	const struct eth_esp32_dev_config *const cfg = dev->config;
+	struct eth_esp32_dev_data *const dev_data = dev->data;
+	int res;
+
+	k_sem_init(&dev_data->int_sem, 1, 1);
+
+	periph_module_enable(PERIPH_EMAC_MODULE);
```

Comment:
I think you should enable ethernet clock using esp32 clock control API.

Reply:
Can you check that I've done this correctly